### ISSUE TITLE
react-i18next: How to sync Browser and Express LanguageDetector?

### ISSUE THREAD

**Issue by acmoune** [i18next/react-i18next#471]
In my server's routes, I do something like this (Note: The` i18next-express-middleware` is already in the pipeline, and Server Side Rendering is working good):

```javascript
export default function (req, res) {
  const lng = req.language.toUpperCase()

  console.log(lng) // Always display [EN]

  MyModel
    .findOne(query, { fieldEN: 1, fieldFR: 1 })
    .exec()
    .then(res => reply(res[`field${lng}`]))
    .catch(err => handle(err))
}
```
So I try to return the right version of the field, base on the USER's selected language.

But no matter what language is selected on the browser side, on the server side it is always set to the default, EN.

Is there a way to let the server side LanguageDetector know about the current language on the browser side ?

This is my `Client i18n init`:

```javascript
import i18n from 'i18next'
import Backend from 'i18next-xhr-backend'
import LanguageDetector from 'i18next-browser-languagedetector'

i18n
  .use(Backend)
  .use(LanguageDetector)
  .init({
    whitelist: ['en', 'fr'],
    fallbackLng: 'en',
    preload: ['en', 'fr'],

    // debug: true,

    interpolation: {
      escapeValue: false
    },

    ns: ['home', 'channel', 'common'],
    defaultNS: 'home',

    backend: {
      loadPath: '/locales/{{lng}}/{{ns}}.json'
    },

    react: {
      wait: true,
      bindI18n: 'languageChanged loaded',
      bindStore: 'added removed',
      nsMode: 'default'
    }
  })

export default i18n
```

And this is my server` i18n init`:

```javascript
import i18n from 'i18next'
import Backend from 'i18next-node-fs-backend'
import { LanguageDetector } from 'i18next-express-middleware'

i18n
  .use(Backend)
  .use(LanguageDetector)
  .init({
    whitelist: ['en', 'fr'],
    fallbackLng: 'en',
    preload: ['en', 'fr'],

    // debug: true,

    interpolation: {
      escapeValue: false
    },

    ns: ['home', 'channel', 'common'],
    defaultNS: 'home',

    backend: {
      loadPath: `${__dirname}/public/locales/{{lng}}/{{ns}}.json`,
      jsonIndent: 2
    }
  })

export default i18n
```

**Comment by jamuhl** [comment]
hm, is your detected language something like `fr-FR`?!? a language-Region combination if so that would not go through the whitelist -> you will need to additional set `nonExplicitWhitelist` on init - https://www.i18next.com/overview/configuration-options#languages-namespaces-resources

just guessing as everything else looks right...

**Comment by acmoune** [comment]
No, it is exactly FR, I set it myself with` i18n.changeLanguage('fr')` from React app.

I tried the `nonExplicitWhitelist` config. It seems to make `req.language` return `en-US` instead of `en`, even if the language is set to `en`. But 

```javascript
console.log(lng) // on top
```
is still returning 'EN-US', instead of 'FR-FR'.

So if I understand you well, both LanguageDetectors (browser and express) should naturally detect the same thing, I think from the browser's LocalStorage ?

**Comment by jamuhl** [comment]
the client detector is not running same time on server...the client pick up from server via set cookie. But on ssr you set it like initialStore via initialLanguage: https://react.i18next.com/components/translate-hoc#the-translate-hoc-props

if i get it right you do something like:

```js
export default function (req, res) {
  const lng = req.language.toUpperCase()

  console.log(lng) // Always display [EN]

  MyModel
    .findOne(query, { fieldEN: 1, fieldFR: 1 })
    .exec()
    .then(res => reply(res[`field${lng}`]))
    .catch(err => handle(err))
}
```

in an express middleware and somewhere inside the react render on server you set language via `i18n.changeLanguage('fr')`...right?

That can't work middleware run before render...language there will be what detector detected if order of middlewares is right - setting it inside rendering is to late.

**Comment by jamuhl** [comment]
you create a i18next instance clone - use that for serverside render. Do you set initialStore, initialLanguage for client?

using an own clone does nothing beside using that for render on server - the instance used on request does not change language but stays on its own language set. req.i18n

**Comment by jamuhl** [comment]
yes that is normal...if detection was en there is no reason to push down fr

**Comment by jamuhl** [comment]
`req.i18n.languages` is the array of languages it uses for the translation lookup....eg. for en-US it would be `en-US` -> `en` -> `fallbackLng` --> in your case with whitelist and fallback en only `[en]`...

if you like to pass down all preloaded you will need to change that to:


```js
const initialI18nStore = {}
  req.i18n.options.preload.forEach((l) => {
    initialI18nStore[l] = req.i18n.services.resourceStore.data[l]
  })
```

but be aware the more languages you add the bigger this gets resulting in pushing all languages to the client...

**Comment by acmoune** [comment]
So `i18n.services.resourceStore.data` keeps all the translations.

On SSR I create the `initialStore` and set the `initialLanguage`. This will be done using the `i18n-server init` config. At this point, I load only the detected language's translations.

Now on the client side, on initial rendering, nothing will be added to the translations store, since the `initialLanguage` translations are there already.

As soon as I `i18n.changeLanguage('another-language')`, the translations store is populated with new translations from the added language.

Am I right up to there ?

Now, on the server side, once `app.use(i18nMiddleware.handle(i18n))` is called, the `req.i18n.language` is set with whatever will be detected as the default language from the` i18n-server` config. So the `initialLanguage` for the client will always be set to that language on SSR, **en** in my case.

If I am still right then my question is: How can I update req.i18n.language according to the changed language from the client, before rendering on the server ? I think I should do it by calling `i18n.changeLanguage(...)` before `app.use(i18nMiddleware.handle(i18n))`, but where will I take the changed language ?

I don't know if the translations store is acting like the Redux store, so whatever update I do on the client store, I can first dispatch the same update to the server store, so on SSR, everything will be in sync.

**Comment by jamuhl** [comment]
First part yes...correct

Second...idk...there is no magic syncing the resourceStore from client to backend. There is nothing doing this for redux neither. On server resourceStore already contains every translation - while to client you only pass what is needed down there.

> If I am still right then my question is: How can I update req.i18n.language according to the changed language from the client, before rendering on the server ?

You mean for a second server side render?

changeLanguage can set cookie on client...next load of a serverside rendered page will pass the cookie up to server which will take that for language detection.

**Comment by acmoune** [comment]
Ok. Thanks for the clarifications. 

Now is there a convention for the cookie, so the server LanguageDetector will autmaticaly take its value, or is there a hook point to push the cookie value to the LanguageDetector,  or should I call `i18n.changeLanguage` before `app.use(i18nMiddleware.handle(i18n))` ?

**Comment by jamuhl** [comment]
for browser set: `caches: ['localStorage', 'cookie']` https://github.com/i18next/i18next-browser-languageDetector#detector-options

same for server: https://github.com/i18next/i18next-express-middleware/blob/master/src/LanguageDetector.js#L17

**Comment by acmoune** [comment]
I am almost done. My only problem now is the language region.
When I set the cookie, with` i18n.changeLanguage(...)`, I have total control on the language format, I set it to either 'en' or 'fr'. Once set, the detector always take from there.

The problem is with the initial detection. The Detector always add the region, and I need to remove it.

This is my actual config, on both side (client and server)

```javascript
...
whitelist: ['en', 'en-US', 'fr', 'fr-FR'],
fallbackLng: 'en',
load: 'languageOnly',
nonExplicitWhitelist: true,
...
```

**Comment by jamuhl** [comment]
detection will always be best match -> so i18next.language will be `en-US` but i18next.languages[0] will be best language you support. (as languageOnly is set it won't load or use `en-US` https://github.com/i18next/i18next/blob/master/src/LanguageUtils.js#L100)

**Comment by acmoune** [comment]
First, thank you so much for your time, I am really grateful.

Finally, everywhere I have to read the language, I just do `i18n.language.substring(0, 2)`, so I get the language only. I think it is a reasonable work around.

I have a last issue (I really hope so), and it seems to be browser related, the worse type, and express related too.

When I read `req.i18n.language `from an API route, with firefox, I get the right one, I mean the one from the cookie. With chrome I always get **en-US**. But in the middleware that does the SSR, both return the good value.

This is part of express setup:

```javascript
app.use(i18nMiddleware)
app.use(middleware.addStoreToRequestPipeline) // Redux store

// API
...
app.get('/api/about', routes.api.about) // req.i18n.language === 'en-US'
...

app.use(middleware.addComponentToRequestPipeline) // inside the middleware, req.i18n.language === 'fr'
app.use(routes.views.index) // The only view
```

I don't know where to start troubleshooting this. I don't know if it is a timing issue, but why is it working with firefox then.
